mqt/cli.py

```diff
@@ -0,0 +1,59 @@
+import os
+import click
+import sys
+
+from .lints import get_count_fails
+from .lints import run_pylint
+from .lints import flake
+
+CLICK_DIR = click.Path(exists=True, dir_okay=True, resolve_path=True)
+
+
+@click.group()
+def cli():
+    """Maintainer Quality tools from OCA helper scripts."""
+    click.echo('Running OCA Maintainer quality tools.!')
+
+
+@cli.command()
```

Comment:
@nhomar IMO each command should be defined in the module where it is implemented  and added as subcommand of a main command. 
In short:
* into setup.py you define one entry point to create a console script for your main command
```python
entry_points='''
    [console_scripts]
    oca-mqt=oca.mqt.cli:main
'''
```

* into cli.py you define the main method with the common args to all the sub commands. You can store the value of common option into the 'ctx' objet. This object is passed to all the sub commands.
```python
@click.group()
@click.version_option(version=__version__, message=__notice__)
@click.option('-v', '--verbose', count=True)
@click.option('-c', '--config', type=click.Path(dir_okay=False, exists=True),
              help="Configuration file (default: ./mqt.cfg).")
@click.pass_context
def main(ctx, verbose, config):
    ctx.obj = config
    if verbose > 1:
        level = logging.DEBUG
    elif verbose > 0:
        level = logging.INFO
    else:
        level = logging.WARNING

    logger = logging.getLogger()
    channel = logging.StreamHandler()
    channel.setFormatter(ColoredFormatter())
    logger.setLevel(level)
    logger.addHandler(channel)
```
* into each module providing sub command, you declare the sub command and add it to the main command.

lint.py
```python
from .cli import main

@click.option('paths', '--path',
              envvar='BUILD_DIR',
              multiple=True,
              type=CLICK_DIR,
              required=True,
              default=[os.getcwd()],
              help="Addons paths to check pylint")
 @click.option('--sys-paths', '-sys-path',
               envvar='PYTHONPATH',
               multiple=True,
               type=CLICK_DIR,
               help="Additional paths to append in sys path.")
 @click.option('--extra-params', '-extra-param',
               multiple=True,
               help="Extra pylint params to append in pylint command")
 @click.option('--msgs-no-count', '-msgs-no-count',
               multiple=True,
               help="List of messages that will not add to the failure count.")
 def lint(paths, ctx, msgs_no_count=None,
          sys_paths=None, extra_params=None):
    """Test the pylint an odoo-addons folder."""
    try:
        fname = cfg.config.name if cfg.config else False
        stats = run_pylint(
            list(paths), cfg=fname,
            sys_paths=sys_paths,
            extra_params=extra_params
        )
        count_fails = get_count_fails(stats, list(msgs_no_count))
    except UserWarning:
        count_fails = -1
    click.echo(count_fails)

main.add_command(lint)
```

In this way you improve the modularity, the readability and the maintainability. 

To run the subcommand lint you just need to type:

```
oca-mqt lint
```

Reply:
I personally prefer only one file with all the cli parts (that little piece of code is yet candidate to be removed).

My statement is that the CLI file for me is like the layer **view** in a MVC architecture.

Then all my cli should not have code with logic, just calling method with parameters, into the index of commands in only 1 place, that point to different Agnostic Python modules.

I think it is easiest to document and to maintain as an unique package.

I do not know if that make sense for you?.

@lmignon (let's discuss what I say and not what I do, remember I am just copying pasting the current code, trying to not refactor anything to move step by step).

About Modularity and plugins I think it is another story.

Reply:
@lmignon Another point about the proposal of architecture I am using is ensure that everything can be importable in a separate script for custom usage (not considered here).

i.e: 

I make my own script without patch mqt itsef that will make:

```
from mqt import some_tool
import click

@click.command
def custom_stuff():
    do_your_own_magic()
    some_tool(**params)
```

Do that make sense for you?

Reply:
@nhomar The problem with this approach is that your 'cli ' must be aware off all the command (a kind of one2many). IMO it's the responsability of a sub commad to register itself into the command registry. In your approach the command registry is responsible to register all the command. 
yours
cli.py
```
from . import x
from . import y
from . import z

def cmd_x():

def cmd_y():

def cmd_z():

```
If you want to add a new command you must modify cli.py and add the module providing the logic

my proposal
cli.py or main.py

```
def main()
```

x.py
```
from .main import main

def MyClassX():
   ...

def my_method_x1():

def cmd_x();

main.add(cmd_x)
```
x is a module that can be used where you want and x register its command into the command registry.

If you want to add a new command  you just have to 'import main ; main.add()' from where you want. (from other module into the oca-mqt package or even from your internal package that leverage oca-mqt for your specific needs)

Reply:
> @nhomar The problem with this approach is that your 'cli ' must be aware off all the command (a kind of one2many). IMO it's the responsability of a sub commad to register itself into the command registry. In your approach the command registry is responsible to register all the command.

@lmignon Thanks for the proper explanation, I think I perfectly understand, well! what you see as an advantage I see it as a problem then we have different PoV there, can I move on this direction as I am doing it now to measure properly the results of my PoV and then if not, it will represent just move 4 lines methods to their proper files which can be done in 5 minutes, Agree?.

It is not necessary to expend a lot of time there now, because neither you and I are **wrong** on my option, On my option we will see only one isolated list of method with TONS of documentation for such option on the command line itself because the point is have a very clean file very well documented in the command line. With yours it will be (in my opinionated PoV) a mess to control that something is properly documented in the cli sub commands. 

Then as technically it will not represent a huge problem once we finish the first version we can change it if **really** necessary. Agree?